### ISSUE TITLE
Closure parameter types are not inferred when variadics are involved

### ISSUE THREAD

**Issue by taminomara** [EmmyLuaLs/emmylua-analyzer-rust#586]
When passing a closure to functions like `pcall`, closure's parameters are not inferred.

Example:

```lua
pcall(function(x)
    local _x = x --- @type string
    --    ^^ should give an error: `x` is not a string
end, 1)
```

Test cases. First two currently don't work, the last one works:

```lua
--- @generic T
--- @param cb fun(...: T...)
--- @param ... T...
function invoke1(cb, ...)
    cb(...)
end

invoke1(
    function(a, b, c)
        local _a = a --- @type integer
        local _b = b --- @type string
        local _c = c --- @type integer
        --    ^^ should give an error: `c` is not an integer
    end,
    1, "2", "3"
)


--- @generic T
--- @param p [T...]
--- @param cb fun(...: T...)
function invoke2(p, cb)
    cb(table.unpack(p))
end

invoke2(
    {1, "2", "3"},
    function(a, b, c)
        local _a = a --- @type integer
        local _b = b --- @type string
        local _c = c --- @type integer
        --    ^^ should give an error: `c` is not an integer
    end
)

--- @generic T1, T2, T3
--- @param p [T1, T2, T3]
--- @param cb fun(_1: T1, _2: T2, _3: T3)
function invoke3(p, cb)
    cb(table.unpack(p))
end

invoke3(
    {1, "2", "3"},
    function(a, b, c)
        local _a = a --- @type integer
        local _b = b --- @type string
        local _c = c --- @type integer
        --    ^^ should give an error: `c` is not an integer
    end
)
```
